### ISSUE TITLE
Dynamic configuration of peers in cbus

### ISSUE THREAD

**Issue by kostja** [tarantool/tarantool#1274]
Rewrite wal_set_watcher()/wal_clear_watcher() to use cbus for messaging, rather than mutex. This way the watcher list won't have to be protected by a mutex.
## Implementaion

Right now there is no bus between relay thread and WAL thread. Manual setting of all peer buses is a shortcoming of the current cbus architecture. Each cord needs to have a local staging area for incoming messages, but if there is no outgoing staging area, it should be possible to push a message directly to the protected input area of the cord.

In other words, the cbus/cpipe structure needs to be refactored, so that cord's protected input area is part of cord structure. Instead of binding mutexes 1:1 to a single exchange, use an array of mutexes and hash cord id to a slot in the array to identify the right mutex to lock.

**Comment by rtsisyk** [comment]
I need this feature in order to implement #723, which is in 1.7.x.  Let's try to fit it into 1.7.x.

**Comment by rtsisyk** [comment]
Implemented.
